node/src/services/blockchain/handler.rs

```diff
@@ -72,19 +207,351 @@ impl ActorEventLoop<BlockchainService> for BlockchainServiceEventLoop {
+                hash == extrinsic_hash
+            })
+            .expect("Extrinsic not found in block. This shouldn't be possible if we're looking into a block for which we got confirmation that the extrinsic was included; qed");
+
+        // Get the events from storage.
+        let events_in_block = self.get_events_storage_element(block_hash)?;
+
+        // Filter the events for the extrinsic.
+        // Each event record is composed of the `phase`, `event` and `topics` fields.
+        // We are interested in those events whose `phase` is equal to `ApplyExtrinsic` with the index of the extrinsic.
+        // For more information see: https://polkadot.js.org/docs/api/cookbook/blocks/#how-do-i-map-extrinsics-to-their-events
+        let events = events_in_block
+            .into_iter()
+            .filter(|ev| ev.phase == frame_system::Phase::ApplyExtrinsic(extrinsic_index as u32))
+            .collect();
+
+        // Construct the extrinsic.
+        Ok(Extrinsic {
+            hash: extrinsic_hash,
+            block_hash,
+            events,
+        })
+    }
+
+    /// Unwatch an extrinsic.
+    async fn unwatch_extrinsic(&self, subscription_id: Number) -> Result<String> {
+        let (result, _rx) = self
+            .rpc_handlers
+            .rpc_query(&format!(
+                r#"{{
+                    "jsonrpc": "2.0",
+                    "method": "author_unwatchExtrinsic",
+                    "params": [{}],
+                    "id": {}
+                }}"#,
+                subscription_id, subscription_id
+            ))
+            .await
+            .expect("Sending query failed even when it is correctly formatted as JSON-RPC; qed");
+
+        let json: serde_json::Value =
+            serde_json::from_str(&result).expect("the result can only be a JSONRPC string; qed");
+        let unwatch_result = json
+            .as_object()
+            .expect("JSON result is always an object; qed")
+            .get("result");
+
+        if let Some(unwatch_result) = unwatch_result {
+            if unwatch_result
+                .as_bool()
+                .expect("Result is always a boolean; qed")
+            {
+                debug!(target: LOG_TARGET, "Extrinsic unwatched successfully");
+            } else {
+                return Err(anyhow::anyhow!("Failed to unwatch extrinsic"));
+            }
+        } else {
+            return Err(anyhow::anyhow!("Failed to unwatch extrinsic"));
+        }
+
+        Ok(result)
+    }
+
+    /// Get the events storage element in a block.
+    fn get_events_storage_element(&self, block_hash: H256) -> Result<EventsVec> {
+        // Would be cool to be able to do this...
+        // let events_storage_key = frame_system::Events::<storage_hub_runtime::Runtime>::hashed_key();
+
+        // Get the events storage key.
+        let events_storage_key = [
+            Twox128::hash(b"System").to_vec(),
+            Twox128::hash(b"Events").to_vec(),
```

Comment:
Do you think we can turn these hashes into consts instead since it is highly improbably that they will ever change?

Reply:
Good suggestion, done